### ISSUE TITLE
Is there interest in new feature: pinning and moving arenas to specific NUMA nodes

### ISSUE THREAD

**Issue by lionkov** [jemalloc/jemalloc#2233]
I work on one of the Exascale Computing Project (https://www.exascaleproject.org) funded projects that uses jemalloc arenas to allocate memory to specific arenas and move arenas from one NUMA node to another. 

We thought that the wider jemalloc community might be interested in that feature. If there is interest, I can create a pull request with the functionality.

**Comment by Lapenkov** [comment]
Hi. Thanks for offering help!
I'm not experienced with NUMA, could you please explain why there might be a need to move arenas between NUMA nodes?   And conceptually, what constitutes the bond between an arena and a NUMA node?

**Comment by lionkov** [comment]
NUMA (Non-uniform memory access) is the most widely used memory architecture from SMP computers. Since each CPU socket has its own memory controllers that it uses to access memory, the memory access to DIMMs connected to that socket is faster than to the DIMMs attached to other CPU sockets. In many cases that difference doesn't matter much, especially with DRAM. But if you have something like Intel's Optane DC NVRAM DIMMs, the memory latency and bandwidth characteristics matter a lot. Additionally, there are architectures that allow GPU memory to be accessible to the CPU as a NUMA node increasing the heterogeneity combinations of different memories. 

So in some cases it is beneficial to create an arena that contains data that will be used together in most of the time and make sure that all arena allocations belong to one of the NUMA domains. Then, at some point of the execution it might make sense to move the data to another NUMA domain. That is part of the functionality our project SICM (https://github.com/lanl/SICM) provides. 

For example, one can have a program that starts running on the CPU, reads/initializes data in an arena, then moves it to the NUMA node that represents GPU memory and continues execution on the GPU. 

I feel that such functionality might be beneficial to other projects, especially if NVRAM and GPU memory access gets more popular.

**Comment by lionkov** [comment]
Any progress on this? Thanks.

**Comment by huntinux** [comment]
> Any progress on this? Thanks.

Interested in this feature. Does arena allocated from specified NUMA node using percpu mode?

**Comment by lionkov** [comment]
No. The idea is that arena can migrate from one NUMA node to another. For example, if you have a CPU/GPU system with unified memory support, you can allocate the arena in the CPU memory, populate the data, and migrate it to the GPU memory.
